### ISSUE TITLE
Remove symfony/http-client dependency

### ISSUE THREAD

**Issue by adamziel** [WordPress/php-toolkit#71]
`symfony/http-client` is a 500kb large dependency that's incompatible with PHP 7.0 and requires the 4,6MB amphp library to handle concurrency when libcurl is missing, and `amphp` is also incompatible with PHP 7.0. Let's replace it. Here's a few ideas:

* Guzzle 4.0
* A custom implementation

All other async HTTP libraries seem to require ReactPHP or amphp.

Related to https://github.com/WordPress/blueprints-library/issues/26

**Comment by adamziel** [comment]
It seems like `fopen("https://...", "r")` call is blocking in that it seems to wait for the initial headers to be received before returning. It can be then switched to a non-blocking mode via `stream_set_blocking`, but at that point we've already lost some blocking time that could have been parallelized. Let's keep digging.

```php

function download_url_fopen( $url ) {
	$contextOptions = [
		'http' => [
			'header' => [
				'Accept: */*',
				'Accept-Encoding: gzip, deflate',
			],
		],
	];
	$context = stream_context_create( $contextOptions );

	$stream = fopen(
		$url,
		'r',
		false,
		$context
	);
	if ( $stream === false ) {
		throw new Exception( 'Unable to open stream' );
	}

	stream_set_blocking( $stream, 0 );

	return $stream;
}
```

**Comment by adamziel** [comment]
For two downloads, `fsockopen()` seems to return `$fp` 5x faster than `fopen()`:

```php
function download_url_fsockopen( $url ) {
	$urlParts = parse_url( $url );
	$host = $urlParts['host'];
	$port = 443;
	$timeout = 30;
	$uri = $urlParts['path'];

	$fp = fsockopen( $host, $port, $errno, $errstr, $timeout );
	if ( ! $fp ) {
		throw new Exception( 'Unable to open stream' );
	}

	return $fp;
}


// Measure time
$time_start = microtime( true );

$fp1 = download_url_fsockopen( 'https://downloads.wordpress.org/plugin/woocommerce.8.6.1.zip' );
echo 'After the first download_url' . PHP_EOL;

$fp2 = download_url_fsockopen( 'https://downloads.wordpress.org/plugin/gutenberg.17.9.0.zip' );
echo 'After the second download_url' . PHP_EOL;

// Measure time
$time_end = microtime( true );
$time = $time_end - $time_start;
echo "Download initiated in $time seconds\n";
```

I'm still not sure whether the socket initiation blocks. The time difference could be explained by not waiting for HTTP response headers (or transmitting the request headers), but it is possible `fsockopen` still waits for the socket to become available.

**Comment by adamziel** [comment]
`stream_socket_client` seems capable of opening connections instantly without blocking. This download function opens two connections instantly (0.001s):

```php
function download_url_stream_socket_client( $url ) {
	$stream = stream_socket_client(
		$url,
		$errno,
		$errstr,
		30,
		STREAM_CLIENT_ASYNC_CONNECT
	);
	if ( $stream === false ) {
		throw new Exception( 'Unable to open stream' );
	}

	stream_set_blocking( $stream, 0 );

	return $stream;
}
```

The challenge is, this is a raw TCP socket without any SSL support. Using `ssl://` as the protocol makes the call slow (0.46s). Can we get both the speed AND lean on the build-in SSL handling?

**Comment by adamziel** [comment]
`stream_socket_client` seems to do the trick and can handle non-blocking SSL sockets thanks to `stream_socket_enable_crypto` and `stream_set_blocking`. Here's a code snippet:

https://gist.github.com/adamziel/c90dd65a3919465e6a03943de792dfe2

I'll move forward with that and let's see how that goes.

**Comment by adamziel** [comment]
A tiny, custom library was implemented in #72. Symfony/http-client is not needed anymore.
